model : add ASR support for LFM2-Audio-1.5B

tdakhran opened the pull request:
[LFM2-Audio-1.5B](https://huggingface.co/LiquidAI/LFM2-Audio-1.5B) supports audio input and audio output.

PR adds only ASR support. To perform ASR invoke CLI with
```
bin/llama-mtmd-cli -m LFM2-Audio-1.5B-F32.gguf --mmproj mmproj-LFM2-Audio-1.5b-F32.gguf -n 30 --audio input.wav -sys "Perform ASR." -p "<__media__>"
```

Changes to existing code:
* model requires system prompt, `-sys` enabled for `llama-mtmd-cli`
* mel bins generation reworked, now it is generated dynamically and supports different `n_fft` values
* `OP_SSM_CONV` for CUDA backend is extended to support kernel size 9

cc: @ngxson

tdakhran commented:
The code is tested, will wait for #17978 to be merged, and then rebase and mark it as "ready for review".

tdakhran committed:
ASR with LFM2-Audio-1.5B

tdakhran committed:
Set rope_theta

CISC commented on convert_hf_to_gguf.py:
```diff
@@ -9560,22 +9563,29 @@ def set_gguf_parameters(self):
         self.gguf_writer.add_vocab_size(self.hparams["vocab_size"])
         self.gguf_writer.add_shortconv_l_cache(self.hparams["conv_L_cache"])
         self.gguf_writer.add_layer_norm_rms_eps(self.hparams["norm_eps"])
+        self.gguf_writer.add_rope_freq_base(self.hparams.get("rope_theta", 1000000))
```
This should not be necessary, but I see why you had to, can you add support for `text_config` fallback here instead?
https://github.com/ggml-org/llama.cpp/blob/5044ab610658f4c8d550e03f1c048f7ff2d1c39b/convert_hf_to_gguf.py#L147-L148

CISC replied:
The problem is that `rope_parameters` doesn't get populated from `text_config`, this needs a fix, then you don't have to `add_rope_freq_base` yourself.

CISC replied:
I'm talking about the init of `rope_parameters`, which is done in the `ModelBase` class, it's not a problem for gguf metadata set in `set_gguf_parameters` as that is in the `TextModel` class, after `text_config` has been merged into `hparams`.

CISC replied:
Fixed in #18061

tdakhran committed:
Fix comment

ngxson commented on tools/mtmd/models/lfm2-audio-enc.cpp:
```diff
@@ -0,0 +1,267 @@
+
+            x = ggml_cont(ctx0, ggml_transpose(ctx0, x));
+
+            // TODO: add support of torch.funtional.glu
```
hmm, can we use `ggml_glu` ?

tdakhran replied:
I tried `ggml_glu`, but it produced numerically different results from `torch.functional.glu`, will give it one more try.

upd: this is sigmoid gating glu, my understanding is that ggml_glu is activation glu that doesn't split the input, please correct me if I wrong.

tdakhran replied:
`ggml_glu` is supposed to split the tensor, gate it, and multiply. Please help me understand how I can express code below differently?
```
                ggml_tensor * gate = ggml_sigmoid(ctx0, ggml_view_2d(ctx0, x, d, x->ne[1], x->nb[1], d * x->nb[0]));
                x                  = ggml_mul(ctx0, ggml_view_2d(ctx0, x, d, x->ne[1], x->nb[1], 0), gate);
```

CISC replied:
You could of course add another glu op for it...

tdakhran replied:
let's keep it as is, implementing extra glu op will require touching many backends, which I'd like to avoid in the scope of this PR

ngxson commented on tools/mtmd/models/lfm2-audio-enc.cpp:
```diff
@@ -0,0 +1,267 @@
+            {
+                x = ggml_cont(ctx0, ggml_transpose(ctx0, x));
+                x = ggml_add(ctx0, ggml_mul(ctx0, x, layer.conv_norm_w), layer.conv_norm_b);
+                x = ggml_cont(ctx0, ggml_transpose(ctx0, x));
```
we may be able to remove of transposes if `conv_norm_b` is already transpose upon conversion?

tdakhran committed:
Remove rope_theta setting

tdakhran commented on tools/mtmd/models/lfm2-audio-enc.cpp:
```diff
@@ -0,0 +1,267 @@
+            auto * conv_pw2_w = ggml_reshape_2d(ctx0, layer.conv_pw2_w, layer.conv_pw2_w->ne[1], layer.conv_pw2_w->ne[2]);
+            x = ggml_mul_mat(ctx0, conv_pw2_w, x);
+            x = ggml_add(ctx0, x, layer.conv_pw2_b);
+            x = ggml_cont(ctx0, ggml_transpose(ctx0, x));
```
Many transposes here are following the Python code without optimization in mind. The objective was to get numerically close intermediates. I'll have a closer look to understand what can be optimized.

tdakhran replied:
removed most of transposes

tdakhran commented on tools/mtmd/models/lfm2-audio-enc.cpp:
```diff
@@ -0,0 +1,267 @@
+    {
+        // layer.0 - conv2d
+        cur = ggml_conv_2d(ctx0, model.pre_encode_conv_X_w[0], cur, 2, 2, 1, 1, 1, 1);
+        cur = ggml_add(ctx0, cur, ggml_reshape_4d(ctx0, model.pre_encode_conv_X_b[0], 1, 1, cur->ne[2], 1));
```
moved reshape into conversion script

tdakhran committed:
Address PR feedback

tdakhran commented:
@ngxson , I addressed most of the feedback, added a comment explaining why `build_attn` cannot be used, removed unnecessary transposes, and simplified permutes. Applied the formatting as well.

PR requires [#18061](https://github.com/ggml-org/llama.cpp/pull/18061), otherwise `rope_theta` won't be set.

tdakhran commented:
Rebased to incorporate https://github.com/ggml-org/llama.cpp/pull/18061, now works as is.

ngxson commented:
@tdakhran hmm, seems like I cannot push to this PR because it was created from an org account

could you re-open another from your personal account?

tdakhran commented:
Hi @elfarolab, and thank you.

GGUFs are not yet updated in https://huggingface.co/LiquidAI/LFM2-Audio-1.5B-GGUF, they are used for [LEAP](https://leap.liquid.ai)l and will be updated together with LEAP.

Use the latest master commit.
Convert the checkpoint https://huggingface.co/LiquidAI/LFM2-Audio-1.5B manually into GGUFs
```
export CKPT=/data/playground/checkpoints/LFM2-Audio-1.5B
python convert_hf_to_gguf.py $CKPT
python convert_hf_to_gguf.py $CKPT --mmproj
```
This will create 2 GGUFs required for ASR
```
❯ (cd $CKPT && ls *.gguf)
LFM2-Audio-1.5B-BF16.gguf  mmproj-LFM2-Audio-1.5b-BF16.gguf
```

Now launch `llama-server` with the command
```
bin/llama-server -m $CKPT/LFM2-Audio-1.5B-BF16.gguf --mmproj $CKPT/mmproj-LFM2-Audio-1.5b-BF16.gguf
```

From another terminal, post a request for ASR.
```
import base64
import pathlib
from openai import OpenAI

wav_file = "/data/playground/issue_400/10.wav"

messages = [
    {"role": "system", "content": "Perform ASR."},
    {
        "role": "user",
        "content": [
            {
                "type": "input_audio",
                "input_audio": {
                    "format": "wav",
                    "data": base64.b64encode(
                        pathlib.Path(wav_file).read_bytes()
                    ).decode("utf-8"),
                },
            },
        ],
    },
]

host = "localhost"
port = 8080
client = OpenAI(base_url=f"http://{host}:{port}/v1", api_key="not-needed")
resp = client.chat.completions.create(
    model="",
    messages=messages,
)
content = resp.choices[0].message.content
print(content)
```

Please let me know if this works.

elfarolab commented:
@tdakhran 

Hello,

I followed the procedure and run the python script above.

---
### The audio file
I used a .WAV 1CH, mono, 16 bits, 24KHz (I know 24KHz looks weird but it is because it will be used later with libopus).
it is composed by a male and female voice, alternating at every sentence, male voice start first. M..F..M..F
Sorry github doesn't allow attaching the .wav audio file.
```
mediainfo speech_orig_24000Hz.wav

General
Complete name                            : speech_orig_24000Hz.wav
Format                                   : Wave
File size                                : 506 KiB
Duration                                 : 10 s 800 ms
Overall bit rate mode                    : Constant
Overall bit rate                         : 384 kb/s
Writing application                      : Lavf62.3.100

Audio
Format                                   : PCM
Format settings                          : Little / Signed
Codec ID                                 : 1
Duration                                 : 10 s 800 ms
Bit rate mode                            : Constant
Bit rate                                 : 384 kb/s
Channel(s)                               : 1 channel
Sampling rate                            : 24.0 kHz
Bit depth                                : 16 bits
Stream size                              : 506 KiB (100%)
```
----

### Results
```
python test_asr_direct_wav_Tarek.py
The birch canoe slid on the smooth planks. "Glue the sheet to the dark blue background", it said. "It's easy to tell the depth of a well. Four hours of steady work faced us."
```

It is almost perfect, it added the extra: 
```
<.. ,it said. ..> 
```
between the female and male voices.

___

llama-server (router mode) debug output:
```
...
Dec 30 14:49:46 xyz llama-server[1717]: [47365] srv  log_server_r: request: GET /models 127.0.0.1 200
Dec 30 14:50:30 xyz llama-server[1717]: srv  proxy_reques: proxying request to model asr on port 55121
Dec 30 14:50:30 xyz llama-server[1717]: [55121] common_chat_params_init_lfm2: Using content relying on the template
Dec 30 14:50:30 xyz llama-server[1717]: [55121] srv  params_from_: Chat format: Content-only
Dec 30 14:50:30 xyz llama-server[1717]: [55121] slot get_availabl: id  1 | task -1 | selected slot by LRU, t_last = -1
Dec 30 14:50:30 xyz llama-server[1717]: [55121] slot launch_slot_: id  1 | task -1 | sampler chain: logits -> penalties -> dry -> top-n-sigma -> top-k -> typical -> top-p -> min-p -> xtc -> temp-ext -> dist
Dec 30 14:50:30 xyz llama-server[1717]: [55121] slot launch_slot_: id  1 | task 0 | processing task
Dec 30 14:50:30 xyz llama-server[1717]: [55121] slot update_slots: id  1 | task 0 | new prompt, n_ctx_slot = 64000, n_keep = 0, task.n_tokens = 155
Dec 30 14:50:30 xyz llama-server[1717]: [55121] slot update_slots: id  1 | task 0 | n_tokens = 0, memory_seq_rm [0, end)
Dec 30 14:50:30 xyz llama-server[1717]: [55121] slot update_slots: id  1 | task 0 | prompt processing progress, n_tokens = 14, batch.n_tokens = 14, progress = 0.090323
Dec 30 14:50:30 xyz llama-server[1717]: [55121] slot update_slots: id  1 | task 0 | n_tokens = 14, memory_seq_rm [14, end)
Dec 30 14:50:30 xyz llama-server[1717]: [55121] srv  process_chun: processing audio...
Dec 30 14:50:30 xyz llama-server[1717]: [55121] encoding audio slice...
Dec 30 14:50:31 xyz llama-server[1717]: [55121] audio slice encoded in 856 ms
Dec 30 14:50:31 xyz llama-server[1717]: [55121] decoding audio batch 1/1, n_tokens_batch = 136
Dec 30 14:50:31 xyz llama-server[1717]: [55121] audio decoded (batch 1/1) in 22 ms
Dec 30 14:50:31 xyz llama-server[1717]: [55121] srv  process_chun: audio processed in 878 ms
Dec 30 14:50:31 xyz llama-server[1717]: [55121] slot update_slots: id  1 | task 0 | prompt processing progress, n_tokens = 155, batch.n_tokens = 5, progress = 1.000000
Dec 30 14:50:31 xyz llama-server[1717]: [55121] slot update_slots: id  1 | task 0 | prompt done, n_tokens = 155, batch.n_tokens = 5
Dec 30 14:50:31 xyz llama-server[1717]: [55121] slot update_slots: id  1 | task 0 | created context checkpoint 1 of 8 (pos_min = 149, pos_max = 149, size = 0.156 MiB)
Dec 30 14:50:32 xyz llama-server[1717]: [55121] slot print_timing: id  1 | task 0 |
Dec 30 14:50:32 xyz llama-server[1717]: [55121] prompt eval time =    1040.77 ms /   155 tokens (    6.71 ms per token,   148.93 tokens per second)
Dec 30 14:50:32 xyz llama-server[1717]: [55121]        eval time =     755.80 ms /    47 tokens (   16.08 ms per token,    62.19 tokens per second)
Dec 30 14:50:32 xyz llama-server[1717]: [55121]       total time =    1796.57 ms /   202 tokens
Dec 30 14:50:32 xyz llama-server[1717]: [55121] slot      release: id  1 | task 0 | stop processing: n_tokens = 201, truncated = 0
Dec 30 14:50:32 xyz llama-server[1717]: [55121] srv  update_slots: all slots are idle
Dec 30 14:50:32 xyz llama-server[1717]: [55121] srv  log_server_r: request: POST /v1/chat/completions 127.0.0.1 200
Dec 30 14:50:32 xyz llama-server[1717]: srv  log_server_r: request: POST /v1/chat/completions 127.0.0.1 200
```

if you need anything else, running tests or try patches, I will be very happy to help.
Thank you so much to everybody.

tdakhran commented:
thanks for testing it @elfarolab , to reduce RAM usage specify `--no-mmap`.

elfarolab commented:
> another possibility is context length, maybe setting `-c 4096` explicitly can help

OK now I've got ~4.2GB, better, with GGUF not yet quantized.

my model.ini:

```
[asr]
load-on-startup = true
m = /opt/usbhd/models/LFM2-Audio-1.5B-GGUF_LiquidAI/LFM2-Audio-1.5B-BF16.gguf
mm = /opt/usbhd/models/LFM2-Audio-1.5B-GGUF_LiquidAI/mmproj-LFM2-Audio-1.5b-BF16.gguf
c = 4096
;threads = 2
temp = 0
;min-p = 0.15
;presence-penalty = 1.05
fit = off
ngl = -1
fa = on
;mlock = on
mmap = off
b = 2048
ub = 2048
```

What about these:
```
temp = 0
;min-p = 0.15
;presence-penalty = 1.05
```

tdakhran commented:
text generation for ASR for LFM2-Audio-1.5B is greedy, params above look good.
